In [27]:
from pathlib import Path


DATA_PATH = Path(r"..\data\raw form\used_cars_data.csv")




In [28]:
import pandas as pd

df = pd.read_csv(DATA_PATH)

print(df.shape)
df.head()

(7253, 14)


,S.No.,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
0,0,Maruti Wagon R LXI CNG,Mumbai,2010,72000,CNG,Manual,First,26.6 km/kg,998 CC,58.16 bhp,5.0,NaN,1.75
1,1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.2 kmpl,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74


In [29]:
print("Dataset shape:", df.shape)
print("\nColumn names:", df.columns.tolist())
print("\nFirst five records:")
display(df.head())

Dataset shape: (7253, 14)

Column names: ['S.No.', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type', 'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats', 'New_Price', 'Price']

First five records:


,S.No.,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
0,0,Maruti Wagon R LXI CNG,Mumbai,2010,72000,CNG,Manual,First,26.6 km/kg,998 CC,58.16 bhp,5.0,NaN,1.75
1,1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.2 kmpl,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74


In [30]:
df.info

<bound method DataFrame.info of       S.No.                                               Name    Location  \
0         0                             Maruti Wagon R LXI CNG      Mumbai   
1         1                   Hyundai Creta 1.6 CRDi SX Option        Pune   
2         2                                       Honda Jazz V     Chennai   
3         3                                  Maruti Ertiga VDI     Chennai   
4         4                    Audi A4 New 2.0 TDI Multitronic  Coimbatore   
...     ...                                                ...         ...   
7248   7248                  Volkswagen Vento Diesel Trendline   Hyderabad   
7249   7249                             Volkswagen Polo GT TSI      Mumbai   
7250   7250                             Nissan Micra Diesel XV     Kolkata   
7251   7251                             Volkswagen Polo GT TSI        Pune   
7252   7252  Mercedes-Benz E-Class 2009-2013 E 220 CDI Avan...       Kochi   

      Year  Kilometers_Driven F

In [31]:
missing_summary = pd.DataFrame({ 
    "missing_count": df.isna().sum(), 
    "missing percentage": (df.isna().sum() / len(df) * 100).round(2)
    })
missing_summary = missing_summary.sort_values(by="missing_count", ascending=False)
missing_summary

,missing_count,missing percentage
New_Price,6247,86.13
Price,1234,17.01
Seats,53,0.73
Engine,46,0.63
Power,46,0.63
Mileage,2,0.03
Fuel_Type,0,0.00
Kilometers_Driven,0,0.00
Year,0,0.00
Location,0,0.00


In [32]:
df.columns = (df.columns.str.strip().str.lower().str.replace(" ", "_").str.replace(".","", regex=False))
print(df.columns.tolist())

['sno', 'name', 'location', 'year', 'kilometers_driven', 'fuel_type', 'transmission', 'owner_type', 'mileage', 'engine', 'power', 'seats', 'new_price', 'price']


In [33]:
df = df.drop(columns=["sno"], errors="ignore")
print(df.columns.tolist())

['name', 'location', 'year', 'kilometers_driven', 'fuel_type', 'transmission', 'owner_type', 'mileage', 'engine', 'power', 'seats', 'new_price', 'price']


In [34]:
import numpy as np

df["brand"] = df["name"].str.split().str[0]
df["mileage_kmpl"] = (
    df["mileage"]
    .astype("string")
    .str.extract(r"([\d.]+)", expand=False)
    .astype(float)
)
df["engine_cc"] = (
    df["engine"]
    .astype("string")
    .str.extract(r"([\d.]+)", expand=False)
    .astype(float)
)
df["power_bhp"] = (
    df["power"]
    .astype("string")
    .str.extract(r"([\d.]+)", expand=False)
    .astype(float)
)
df["price"] = pd.to_numeric(df["price"], errors="coerce")

REFERENCE_YEAR = 2020
df["vehicle_age"] = REFERENCE_YEAR - df["year"]
display(
    df[
        [
            "name",
            "brand",
            "year",
            "vehicle_age",
            "mileage",
            "mileage_kmpl",
            "engine",
            "engine_cc",
            "power",
            "power_bhp",
            "seats",
            "price",
        ]
    ].head(10)
)

,name,brand,year,vehicle_age,mileage,mileage_kmpl,engine,engine_cc,power,power_bhp,seats,price
0,Maruti Wagon R LXI CNG,Maruti,2010,10,26.6 km/kg,26.60,998 CC,998.0,58.16 bhp,58.16,5.0,1.75
1,Hyundai Creta 1.6 CRDi SX Option,Hyundai,2015,5,19.67 kmpl,19.67,1582 CC,1582.0,126.2 bhp,126.20,5.0,12.50
2,Honda Jazz V,Honda,2011,9,18.2 kmpl,18.20,1199 CC,1199.0,88.7 bhp,88.70,5.0,4.50
3,Maruti Ertiga VDI,Maruti,2012,8,20.77 kmpl,20.77,1248 CC,1248.0,88.76 bhp,88.76,7.0,6.00
4,Audi A4 New 2.0 TDI Multitronic,Audi,2013,7,15.2 kmpl,15.20,1968 CC,1968.0,140.8 bhp,140.80,5.0,17.74
5,Hyundai EON LPG Era Plus Option,Hyundai,2012,8,21.1 km/kg,21.10,814 CC,814.0,55.2 bhp,55.20,5.0,2.35
6,Nissan Micra Diesel XV,Nissan,2013,7,23.08 kmpl,23.08,1461 CC,1461.0,63.1 bhp,63.10,5.0,3.50
7,Toyota Innova Crysta 2.8 GX AT 8S,Toyota,2016,4,11.36 kmpl,11.36,2755 CC,2755.0,171.5 bhp,171.50,8.0,17.50
8,Volkswagen Vento Diesel Comfortline,Volkswagen,2013,7,20.54 kmpl,20.54,1598 CC,1598.0,103.6 bhp,103.60,5.0,5.20
9,Tata Indica Vista Quadrajet LS,Tata,2012,8,22.3 kmpl,22.30,1248 CC,1248.0,74 bhp,74.00,5.0,1.95


In [35]:
df_train = df.dropna(subset=["price"]).copy()
print("Raw dataset rows:", len(df))
print("Rows with known prices:", len(df_train))
print("Rows excluded because price is missing:", len(df) - len(df_train))

Raw dataset rows: 7253
Rows with known prices: 6019
Rows excluded because price is missing: 1234


In [36]:
df_unpriced = df[df["price"].isna()].copy()
print("Unpriced records:", len(df_unpriced))

Unpriced records: 1234


In [37]:
zero_mileage = df_train[df_train["mileage_kmpl"] <= 0]
print("Zero or negative mileage records:", len(zero_mileage))

display(
    zero_mileage[
        [
            "name",
            "year",
            "fuel_type",
            "mileage",
            "mileage_kmpl",
            "engine_cc",
            "power_bhp",
            "price",
        ]
    ].head(20)
)

Zero or negative mileage records: 68


,name,year,fuel_type,mileage,mileage_kmpl,engine_cc,power_bhp,price
14,Land Rover Freelander 2 TD4 SE,2012,Diesel,0.0 kmpl,0.0,2179.0,115.00,17.50
67,Mercedes-Benz C-Class Progressive C 220d,2019,Diesel,0.0 kmpl,0.0,1950.0,194.00,35.67
79,Hyundai Santro Xing XL,2005,Petrol,0.0 kmpl,0.0,1086.0,NaN,1.30
194,Honda City 1.5 GXI,2007,Petrol,0.0 kmpl,0.0,NaN,NaN,2.95
229,Ford Figo Diesel,2015,Diesel,0.0 kmpl,0.0,1498.0,99.00,3.60
262,Hyundai Santro Xing XL,2006,Petrol,0.0 kmpl,0.0,1086.0,NaN,1.75
307,Hyundai Santro Xing XL,2006,Petrol,0.0 kmpl,0.0,1086.0,NaN,1.50
424,Volkswagen Jetta 2007-2011 1.9 L TDI,2010,Diesel,0.0 kmpl,0.0,1968.0,NaN,5.45
443,Hyundai Santro GLS I - Euro I,2012,Petrol,0.0 kmpl,0.0,1086.0,NaN,3.35
544,Mercedes-Benz New C-Class Progressive C 200,2019,Petrol,0.0 kmpl,0.0,1950.0,181.43,38.99


In [38]:
df_train.loc[df_train["mileage_kmpl"] <= 0, "mileage_kmpl"] = np.nan

In [39]:
numeric_columns = [
    "year",
    "kilometers_driven",
    "mileage_kmpl",
    "engine_cc",
    "power_bhp",
    "seats",
    "vehicle_age",
    "price",
]

df_train[numeric_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
year,6019.0,2013.358199,3.269742,1998.00,2011.0,2014.00,2016.00,2019.00
kilometers_driven,6019.0,58738.380296,91268.843206,171.00,34000.0,53000.00,73000.00,6500000.00
mileage_kmpl,5949.0,18.342252,4.175475,6.40,15.3,18.20,21.10,33.54
engine_cc,5983.0,1621.276450,601.355233,72.00,1198.0,1493.00,1984.00,5998.00
power_bhp,5876.0,113.253050,53.874957,34.20,75.0,97.70,138.10,560.00
seats,5977.0,5.278735,0.808840,0.00,5.0,5.00,5.00,10.00
vehicle_age,6019.0,6.641801,3.269742,1.00,4.0,6.00,9.00,22.00
price,6019.0,9.479468,11.187917,0.44,3.5,5.64,9.95,160.00


In [40]:
quality_checks = pd.Series({
    "Training rows": len(df_train),
    "Rows missing mileage_kmpl": df_train["mileage_kmpl"].isna().sum(),
    "Rows missing engine_cc": df_train["engine_cc"].isna().sum(),
    "Rows missing power_bhp": df_train["power_bhp"].isna().sum(),
    "Rows missing seats": df_train["seats"].isna().sum(),
    "Zero or negative mileage": (df_train["mileage_kmpl"] <= 0).sum(),
    "Zero or negative price": (df_train["price"] <= 0).sum(),
    "Negative vehicle age": (df_train["vehicle_age"] < 0).sum(),
    "Exact duplicate rows": df_train.duplicated().sum(),
}).sort_values(ascending=False)

quality_checks

Training rows                6019
Rows missing power_bhp        143
Rows missing mileage_kmpl      70
Rows missing seats             42
Rows missing engine_cc         36
Zero or negative mileage        0
Zero or negative price          0
Negative vehicle age            0
Exact duplicate rows            0
dtype: int64

In [41]:
import numpy as np

# 1. Treat zero or negative mileage as invalid/missing.
df_train.loc[df_train["mileage_kmpl"] <= 0, "mileage_kmpl"] = np.nan

# 2. A listed vehicle cannot have zero or negative seats.
df_train.loc[df_train["seats"] <= 0, "seats"] = np.nan

# 3. Engine displacement below 500 cc is not credible for this dataset.
# Inspect the affected records before keeping this threshold permanently.
df_train.loc[df_train["engine_cc"] < 500, "engine_cc"] = np.nan

# 4. Kilometres driven above 500,000 are treated as likely data-entry errors.
# Preserve the record but remove the implausible odometer value.
df_train.loc[
    df_train["kilometers_driven"] > 500_000,
    "kilometers_driven"
] = np.nan

In [42]:
display(
    df_train.loc[
        df_train["kilometers_driven"].isna(),
        [
            "name",
            "year",
            "kilometers_driven",
            "fuel_type",
            "transmission",
            "price"
        ]
    ]
)

,name,year,kilometers_driven,fuel_type,transmission,price
340,Skoda Octavia Ambition Plus 2.0 TDI AT,2013,NaN,Diesel,Automatic,7.5
358,Hyundai i10 Magna 1.2,2009,NaN,Petrol,Manual,2.7
1860,Volkswagen Vento Diesel Highline,2013,NaN,Diesel,Manual,5.9
2328,BMW X5 xDrive 30d M Sport,2017,NaN,Diesel,Automatic,65.0


In [43]:
df_train["high_kilometers_flag"] = (
    df_train["kilometers_driven"] > 500_000
)

display(
    df_train.loc[
        df_train["high_kilometers_flag"],
        [
            "name",
            "year",
            "kilometers_driven",
            "fuel_type",
            "transmission",
            "price"
        ]
    ]
)

,name,year,kilometers_driven,fuel_type,transmission,price


In [44]:
df_train.loc[
    df_train["high_kilometers_flag"],
    "kilometers_driven"
] = np.nan

In [45]:
display(
    df_train.loc[
        df_train["engine_cc"].notna() & (df_train["engine_cc"] < 500),
        [
            "name",
            "year",
            "engine",
            "engine_cc",
            "fuel_type",
            "price"
        ]
    ]
)

,name,year,engine,engine_cc,fuel_type,price


In [46]:
df_train["low_engine_flag"] = (
    df_train["engine_cc"].notna()
    & (df_train["engine_cc"] < 500)
)

df_train.loc[df_train["low_engine_flag"], "engine_cc"] = np.nan

In [47]:
display(
    df_train.loc[
        df_train["seats"].notna() & (df_train["seats"] <= 0),
        [
            "name",
            "year",
            "seats",
            "price"
        ]
    ]
)

,name,year,seats,price


In [48]:
df_train["invalid_seats_flag"] = (
    df_train["seats"].notna()
    & (df_train["seats"] <= 0)
)

df_train.loc[df_train["invalid_seats_flag"], "seats"] = np.nan

In [49]:
model_columns = [
    "name",
    "brand",
    "location",
    "year",
    "vehicle_age",
    "kilometers_driven",
    "fuel_type",
    "transmission",
    "owner_type",
    "mileage_kmpl",
    "engine_cc",
    "power_bhp",
    "seats",
    "price",
]

df_model = df_train[model_columns].copy()

print("Final modelling dataset shape:", df_model.shape)

display(df_model.head())

print("\nRemaining missing values:")
display(
    df_model.isna()
    .sum()
    .sort_values(ascending=False)
)

Final modelling dataset shape: (6019, 14)


,name,brand,location,year,vehicle_age,kilometers_driven,fuel_type,transmission,owner_type,mileage_kmpl,engine_cc,power_bhp,seats,price
0,Maruti Wagon R LXI CNG,Maruti,Mumbai,2010,10,72000.0,CNG,Manual,First,26.60,998.0,58.16,5.0,1.75
1,Hyundai Creta 1.6 CRDi SX Option,Hyundai,Pune,2015,5,41000.0,Diesel,Manual,First,19.67,1582.0,126.20,5.0,12.50
2,Honda Jazz V,Honda,Chennai,2011,9,46000.0,Petrol,Manual,First,18.20,1199.0,88.70,5.0,4.50
3,Maruti Ertiga VDI,Maruti,Chennai,2012,8,87000.0,Diesel,Manual,First,20.77,1248.0,88.76,7.0,6.00
4,Audi A4 New 2.0 TDI Multitronic,Audi,Coimbatore,2013,7,40670.0,Diesel,Automatic,Second,15.20,1968.0,140.80,5.0,17.74



Remaining missing values:


power_bhp            143
mileage_kmpl          70
seats                 43
engine_cc             37
kilometers_driven      4
name                   0
vehicle_age            0
year                   0
location               0
brand                  0
owner_type             0
transmission           0
fuel_type              0
price                  0
dtype: int64

## Data-cleaning decisions

- Removed the serial-number field because it is an identifier, not a vehicle attribute.
- Excluded `new_price` from modelling because 86.13% of values were missing.
- Excluded 1,234 observations with missing `price` only from supervised model training because price is the prediction target.
- Extracted numeric values from mileage, engine, and power fields.
- Converted 0.0 kmpl mileage values to missing because they are implausible and appear to encode unavailable data.
- Replaced invalid seat counts and implausible engine/odometer values with missing values rather than deleting the full listing.
- Deferred numeric missing-value imputation until model training, where it will be fitted only on training data.

In [50]:
from pathlib import Path

OUTPUT_PATH = Path(r"..\data\processed form\used_cars_cleaned.csv")
OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)

df_model.to_csv(OUTPUT_PATH, index=False)

print(f"Saved cleaned data to: {OUTPUT_PATH.resolve()}")

Saved cleaned data to: C:\Users\kenda\Documents\Projects\used-car-price-prediction\data\processed form\used_cars_cleaned.csv
